src/cpp/redis.cpp

```diff
@@ -329,29 +329,50 @@ CommandReply Redis::get_script(const std::string& key)
 
 inline void Redis::_connect(std::string address_port)
 {
+    // Note that this logic flow differs from a cluster
+    // because the non-cluster Redis constructor
+    // does not form a connection until a command is run
+
     this->_address_node_map.insert({address_port, nullptr});
 
-    int n_connection_trials = 10;
+    // Try to create the sw::redis::Redis object
+    try {
+        this->_redis = new sw::redis::Redis(address_port);
+    }
+    catch (std::bad_alloc& e) {
+        throw std::runtime_error(e.what());
+    }
+    catch (std::exception& e) {
+        delete this->_redis;
+        this->_redis = 0;
+        throw std::runtime_error("Failed to create Redis "\
+                                 "object with error: " +
+                                 std::string(e.what()));
+    }
+
+    // Attempt to have the sw::redis::Redis object
+    // make a connection using the PING command
+    int n_trials = 10;
 
-    while(n_connection_trials > 0) {
+    for(int i=1; i<=n_trials; i++) {
         try {
-            this->_redis = new sw::redis::Redis(address_port);
-            n_connection_trials = -1;
+            if(this->_redis->ping().compare("PONG")==0) {
+                break;
+            }
         }
-        catch (sw::redis::TimeoutError &e) {
-          std::cout << "WARNING: Caught redis TimeoutError: "
-                    << e.what() << std::endl;
-          std::cout << "WARNING: TimeoutError occurred with "\
-                       "initial client connection.";
-          std::cout << "WARNING: "<< n_connection_trials
-                      << " more trials will be made.";
-          n_connection_trials--;
-          std::this_thread::sleep_for(std::chrono::seconds(2));
+        catch (sw::redis::Error& e) {
+            if(i == n_trials) {
+                delete this->_redis;
+                this->_redis = 0;
+                throw std::runtime_error(e.what());
+            }
+            std::this_thread::sleep_for(std::chrono::seconds(2));
```

Comment:
Sleeping from inside a catch block is dangerous -- if the sleep call throws an exception, there is no way to catch it and the app shuts down immediately.

Reply:
Just set a failed boolean and then do the entire if block outside the exception handler

Reply:
Yeah, I can move it outside if we wanted to be extra safe, but ``sleep_for`` has a no-throw guarantee when using durations defined in std::chrono.